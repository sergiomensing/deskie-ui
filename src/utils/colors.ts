export const colorShade = (col: string, amt: number) => {
  col = col.replace(/^#/, '');
  if (col.length === 3) col = col[0] + col[0] + col[1] + col[1] + col[2] + col[2];

  const [sr, sg, sb] = col.match(/.{2}/g) || ['ff', 'ff', 'ff'];
  const [ar, ag, ab] = [parseInt(sr, 16) + amt, parseInt(sg, 16) + amt, parseInt(sb, 16) + amt];

  const r = Math.max(Math.min(255, ar), 0).toString(16);
  const g = Math.max(Math.min(255, ag), 0).toString(16);
  const b = Math.max(Math.min(255, ab), 0).toString(16);

  const rr = (r.length < 2 ? '0' : '') + r;
  const gg = (g.length < 2 ? '0' : '') + g;
  const bb = (b.length < 2 ? '0' : '') + b;

  return `#${rr}${gg}${bb}`;
};

const defaultColors = [
  '#E64980',
  '#4C6EF5',
  '#228BE6',
  '#15AABF',
  '#12B886',
  '#40C057',
  '#82C91E',
  '#FAB005',
  '#FD7E14',
  '#FA5252',
];

export function getColorFromString(value: string, colors: undefined | string[] = defaultColors) {
  let hash = 0;
  if (value.length === 0) return colors[hash];
  for (let i = 0; i < value.length; i++) {
    hash = value.charCodeAt(i) + ((hash << 5) - hash);
    hash = hash & hash;
  }
  hash = ((hash % colors.length) + colors.length) % colors.length;
  return colors[hash];
}
