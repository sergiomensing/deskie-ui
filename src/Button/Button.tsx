import { PropsWithChildren, ButtonHTMLAttributes, ReactNode } from 'react';
import styled, { css, DefaultTheme, FlattenInterpolation, ThemeProps } from 'styled-components';
import { colorShade } from '../utils/colors';

type ButtonSize = 'default' | 'small';
type ButtonAppearance = 'default' | 'secondary' | 'danger';

type ButtonProps = ButtonHTMLAttributes<unknown> & {
  size?: ButtonSize;
  appearance?: ButtonAppearance;
  loading?: boolean;
  icon?: ReactNode;
  iconRight?: ReactNode;
};

type CssPartial = FlattenInterpolation<ThemeProps<DefaultTheme>>;

const StyledButton = styled.button<{ size: CssPartial; colors: CssPartial; disabled: boolean; iconOnly: boolean }>`
  border: 1px solid transparent;
  font: inherit;
  ${(props) => props.size};
  ${(props) => props.colors};
  display: inline-flex;
  font-weight: 500;
  font-size: 0.875rem;
  user-select: none;
  text-shadow: none;
  justify-content: center;
  align-items: center;
  text-align: center;
  white-space: nowrap;
  transition: background-color 200ms ease 0ms, box-shadow 200ms ease 0ms, border 200ms ease 0ms, color 200ms ease 0ms;
  position: relative;
  overflow: hidden;
  cursor: ${(props) => (props.disabled ? 'not-allowed' : 'pointer')};

  .button-icon {
    width: 1.25rem;

    svg {
      width: 100%;
      height: auto;
    }
  }

  ${(props) =>
    !props.iconOnly &&
    css`
      .button-icon {
        margin-right: 0.5rem;
        margin-left: -0.25rem;
      }
      .button-icon-right {
        margin-right: -0.25rem;
        margin-left: 0.5rem;
      }
    `}
`;

function Button({
  className = '',
  appearance = 'default',
  size = 'default',
  disabled = false,
  icon,
  iconRight,
  children,
  ...rest
}: PropsWithChildren<ButtonProps>) {
  return (
    <StyledButton
      disabled={disabled}
      colors={getColors(appearance, disabled)}
      className={`button ${children ? '' : 'button-iconOnly'} ${className}`}
      iconOnly={!children}
      size={getButtonSize(size, !children)}
      {...rest}
    >
      {icon && <span className="button-icon">{icon}</span>}
      {children && children}
      {iconRight && <span className="button-icon button-icon-right">{iconRight}</span>}
    </StyledButton>
  );
}

function getColors(appearance: ButtonAppearance, disabled: boolean) {
  if (disabled) {
    return css`
      background-color: ${(props) => props.theme.colors.disabled};
      border-color: ${(props) => props.theme.colors.disabled};
      color: ${(props) => props.theme.colors.disabledText};
    `;
  }

  if (appearance === 'secondary') {
    return css`
      background-color: transparent;
      border-color: transparent;
      color: ${(props) => props.theme.colors.black};

      &:hover {
        background-color: ${(props) => props.theme.colors.accent1};
        border-color: ${(props) => props.theme.colors.accent1};
      }
    `;
  }

  if (appearance === 'danger') {
    return css`
      background-color: ${(props) => props.theme.colors.error};
      border-color: ${(props) => props.theme.colors.error};
      color: ${(props) => props.theme.colors.white};

      &:hover {
        background-color: ${(props) => colorShade(props.theme.colors.error, -30)};
        border-color: ${(props) => colorShade(props.theme.colors.error, -30)};
      }
    `;
  }

  return css`
    background-color: ${(props) => props.theme.colors.primary};
    border-color: ${(props) => props.theme.colors.primary};
    color: ${(props) => props.theme.colors.white};

    &:hover {
      background-color: ${(props) => props.theme.colors.primaryHover};
      border-color: ${(props) => props.theme.colors.primaryHover};
    }
  `;
}

function getButtonSize(size: ButtonSize, iconOnly: boolean) {
  if (size === 'small') {
    return css`
      height: 2.5rem;
      border-radius: 0.75rem;
      padding: 0 0.75rem;

      ${iconOnly && 'width: 2.5rem;'}
    `;
  }

  return css`
    height: 3.5rem;
    border-radius: 1rem;
    ${!iconOnly && 'padding: 0 1.25rem;'}
    ${iconOnly && 'width: 3.5rem;'}
  `;
}

export default Button;
