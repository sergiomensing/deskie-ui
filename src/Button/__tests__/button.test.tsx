import React from 'react';
import Button from '../Button';
import { render } from '../../../tests/utils';

describe('Button', () => {
  it('should support all appearances', () => {
    const { container } = render(
      <>
        <Button appearance="default">test-value</Button>
        <Button appearance="secondary">test-value</Button>
        <Button appearance="danger">test-value</Button>
      </>,
    );
    expect(container).toMatchSnapshot();
  });

  it('should render icons', () => {
    const { getByTestId, container } = render(
      <>
        <Button icon={<span data-testid="test-icon"></span>}>test-value</Button>
        <Button iconRight={<span data-testid="test-icon-right"></span>}>test-value</Button>
      </>,
    );

    expect(getByTestId('test-icon')).toBeInTheDocument();
    expect(getByTestId('test-icon-right')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('should support all sizes', () => {
    const { container } = render(
      <>
        <Button size="default">test-value</Button>
        <Button size="small">test-value</Button>
      </>,
    );
    expect(container).toMatchSnapshot();
  });

  it('should render as "Icon Button" when there are no children but there is an icon', () => {
    const { getByTestId } = render(<Button data-testid="test-id" icon={<span data-testid="test-icon"></span>} />);
    expect(getByTestId('test-id')).toMatchSnapshot();
  });
});
