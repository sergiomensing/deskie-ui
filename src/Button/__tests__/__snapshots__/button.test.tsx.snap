// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render as "Icon Button" when there are no children but there is an icon 1`] = `
.c0 {
  border: 1px solid transparent;
  font: inherit;
  height: 3.5rem;
  border-radius: 1rem;
  width: 3.5rem;
  background-color: #6C5DD3;
  border-color: #6C5DD3;
  color: #FFFFFF;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-weight: 500;
  font-size: 0.875rem;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  text-shadow: none;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  white-space: nowrap;
  -webkit-transition: background-color 200ms ease 0ms,box-shadow 200ms ease 0ms,border 200ms ease 0ms,color 200ms ease 0ms;
  transition: background-color 200ms ease 0ms,box-shadow 200ms ease 0ms,border 200ms ease 0ms,color 200ms ease 0ms;
  position: relative;
  overflow: hidden;
  cursor: pointer;
}

.c0:hover {
  background-color: #4535B7;
  border-color: #4535B7;
}

.c0 .button-icon {
  width: 1.25rem;
}

.c0 .button-icon svg {
  width: 100%;
  height: auto;
}

<button
  class="c0 button button-iconOnly "
  data-testid="test-id"
>
  <span
    class="button-icon"
  >
    <span
      data-testid="test-icon"
    />
  </span>
</button>
`;

exports[`Button should render icons 1`] = `
.c0 {
  border: 1px solid transparent;
  font: inherit;
  height: 3.5rem;
  border-radius: 1rem;
  padding: 0 1.25rem;
  background-color: #6C5DD3;
  border-color: #6C5DD3;
  color: #FFFFFF;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-weight: 500;
  font-size: 0.875rem;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  text-shadow: none;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  white-space: nowrap;
  -webkit-transition: background-color 200ms ease 0ms,box-shadow 200ms ease 0ms,border 200ms ease 0ms,color 200ms ease 0ms;
  transition: background-color 200ms ease 0ms,box-shadow 200ms ease 0ms,border 200ms ease 0ms,color 200ms ease 0ms;
  position: relative;
  overflow: hidden;
  cursor: pointer;
}

.c0:hover {
  background-color: #4535B7;
  border-color: #4535B7;
}

.c0 .button-icon {
  width: 1.25rem;
}

.c0 .button-icon svg {
  width: 100%;
  height: auto;
}

.c0 .button-icon {
  margin-right: 0.5rem;
  margin-left: -0.25rem;
}

.c0 .button-icon-right {
  margin-right: -0.25rem;
  margin-left: 0.5rem;
}

<div>
  <button
    class="c0 button  "
  >
    <span
      class="button-icon"
    >
      <span
        data-testid="test-icon"
      />
    </span>
    test-value
  </button>
  <button
    class="c0 button  "
  >
    test-value
    <span
      class="button-icon button-icon-right"
    >
      <span
        data-testid="test-icon-right"
      />
    </span>
  </button>
</div>
`;

exports[`Button should support all appearances 1`] = `
.c0 {
  border: 1px solid transparent;
  font: inherit;
  height: 3.5rem;
  border-radius: 1rem;
  padding: 0 1.25rem;
  background-color: #6C5DD3;
  border-color: #6C5DD3;
  color: #FFFFFF;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-weight: 500;
  font-size: 0.875rem;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  text-shadow: none;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  white-space: nowrap;
  -webkit-transition: background-color 200ms ease 0ms,box-shadow 200ms ease 0ms,border 200ms ease 0ms,color 200ms ease 0ms;
  transition: background-color 200ms ease 0ms,box-shadow 200ms ease 0ms,border 200ms ease 0ms,color 200ms ease 0ms;
  position: relative;
  overflow: hidden;
  cursor: pointer;
}

.c0:hover {
  background-color: #4535B7;
  border-color: #4535B7;
}

.c0 .button-icon {
  width: 1.25rem;
}

.c0 .button-icon svg {
  width: 100%;
  height: auto;
}

.c0 .button-icon {
  margin-right: 0.5rem;
  margin-left: -0.25rem;
}

.c0 .button-icon-right {
  margin-right: -0.25rem;
  margin-left: 0.5rem;
}

.c1 {
  border: 1px solid transparent;
  font: inherit;
  height: 3.5rem;
  border-radius: 1rem;
  padding: 0 1.25rem;
  background-color: transparent;
  border-color: transparent;
  color: #11142D;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-weight: 500;
  font-size: 0.875rem;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  text-shadow: none;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  white-space: nowrap;
  -webkit-transition: background-color 200ms ease 0ms,box-shadow 200ms ease 0ms,border 200ms ease 0ms,color 200ms ease 0ms;
  transition: background-color 200ms ease 0ms,box-shadow 200ms ease 0ms,border 200ms ease 0ms,color 200ms ease 0ms;
  position: relative;
  overflow: hidden;
  cursor: pointer;
}

.c1:hover {
  background-color: #F7F7FD;
  border-color: #F7F7FD;
}

.c1 .button-icon {
  width: 1.25rem;
}

.c1 .button-icon svg {
  width: 100%;
  height: auto;
}

.c1 .button-icon {
  margin-right: 0.5rem;
  margin-left: -0.25rem;
}

.c1 .button-icon-right {
  margin-right: -0.25rem;
  margin-left: 0.5rem;
}

.c2 {
  border: 1px solid transparent;
  font: inherit;
  height: 3.5rem;
  border-radius: 1rem;
  padding: 0 1.25rem;
  background-color: #FF754C;
  border-color: #FF754C;
  color: #FFFFFF;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-weight: 500;
  font-size: 0.875rem;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  text-shadow: none;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  white-space: nowrap;
  -webkit-transition: background-color 200ms ease 0ms,box-shadow 200ms ease 0ms,border 200ms ease 0ms,color 200ms ease 0ms;
  transition: background-color 200ms ease 0ms,box-shadow 200ms ease 0ms,border 200ms ease 0ms,color 200ms ease 0ms;
  position: relative;
  overflow: hidden;
  cursor: pointer;
}

.c2:hover {
  background-color: #e1572e;
  border-color: #e1572e;
}

.c2 .button-icon {
  width: 1.25rem;
}

.c2 .button-icon svg {
  width: 100%;
  height: auto;
}

.c2 .button-icon {
  margin-right: 0.5rem;
  margin-left: -0.25rem;
}

.c2 .button-icon-right {
  margin-right: -0.25rem;
  margin-left: 0.5rem;
}

<div>
  <button
    class="c0 button  "
  >
    test-value
  </button>
  <button
    class="c1 button  "
  >
    test-value
  </button>
  <button
    class="c2 button  "
  >
    test-value
  </button>
</div>
`;

exports[`Button should support all sizes 1`] = `
.c0 {
  border: 1px solid transparent;
  font: inherit;
  height: 3.5rem;
  border-radius: 1rem;
  padding: 0 1.25rem;
  background-color: #6C5DD3;
  border-color: #6C5DD3;
  color: #FFFFFF;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-weight: 500;
  font-size: 0.875rem;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  text-shadow: none;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  white-space: nowrap;
  -webkit-transition: background-color 200ms ease 0ms,box-shadow 200ms ease 0ms,border 200ms ease 0ms,color 200ms ease 0ms;
  transition: background-color 200ms ease 0ms,box-shadow 200ms ease 0ms,border 200ms ease 0ms,color 200ms ease 0ms;
  position: relative;
  overflow: hidden;
  cursor: pointer;
}

.c0:hover {
  background-color: #4535B7;
  border-color: #4535B7;
}

.c0 .button-icon {
  width: 1.25rem;
}

.c0 .button-icon svg {
  width: 100%;
  height: auto;
}

.c0 .button-icon {
  margin-right: 0.5rem;
  margin-left: -0.25rem;
}

.c0 .button-icon-right {
  margin-right: -0.25rem;
  margin-left: 0.5rem;
}

.c1 {
  border: 1px solid transparent;
  font: inherit;
  height: 2.5rem;
  border-radius: 0.75rem;
  padding: 0 0.75rem;
  background-color: #6C5DD3;
  border-color: #6C5DD3;
  color: #FFFFFF;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-weight: 500;
  font-size: 0.875rem;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  text-shadow: none;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  white-space: nowrap;
  -webkit-transition: background-color 200ms ease 0ms,box-shadow 200ms ease 0ms,border 200ms ease 0ms,color 200ms ease 0ms;
  transition: background-color 200ms ease 0ms,box-shadow 200ms ease 0ms,border 200ms ease 0ms,color 200ms ease 0ms;
  position: relative;
  overflow: hidden;
  cursor: pointer;
}

.c1:hover {
  background-color: #4535B7;
  border-color: #4535B7;
}

.c1 .button-icon {
  width: 1.25rem;
}

.c1 .button-icon svg {
  width: 100%;
  height: auto;
}

.c1 .button-icon {
  margin-right: 0.5rem;
  margin-left: -0.25rem;
}

.c1 .button-icon-right {
  margin-right: -0.25rem;
  margin-left: 0.5rem;
}

<div>
  <button
    class="c0 button  "
  >
    test-value
  </button>
  <button
    class="c1 button  "
  >
    test-value
  </button>
</div>
`;
